### PR TITLE
feat: CLI Tools provider API implementation for tools registration

### PR DESCRIPTION
**Pull request by dgolovin**
WIP implementation to let an extension to declare required CLI tools.

### What does this PR do?

Adds extension point, simplified API implementation, UI implementation and extension API call from docker-compose.

### Screenshot/screencast of this PR

https://github.com/containers/podman-desktop/assets/620330/0840e5d4-62ac-48b2-bae4-e0ec7fde31b3

https://github.com/containers/podman-desktop/assets/620330/4afe850d-b4aa-425a-a1d9-85ccc24ef183

### What issues does this PR fix or reference?

#4120

Dependency of 
* https://github.com/containers/podman-desktop/pull/4387
* https://github.com/containers/podman-desktop/pull/4388

### CODE REVIEW

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts]: I think we should go with the `registerCliToolProvider` approach and providing a `CliTool`
then we can return a Disposable

if we have a `create***()` method, an object should be returned and not a Disposable (return CliTool object)

but it looks more to be a `provider` thing to me

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts:2437]: this type is not defined in the .d.ts file

**Comment by dgolovin** [comment]
@benoitf This PR is seems to be pursuing the same goal as https://github.com/containers/podman-desktop/pull/4092.

Some entities names and UI implementation are different, but goals are the same. 
Should we agree on how we name the first class objects and just merge the efforts.

We can basically abandon CliToolsRegistry and use Binary Registry form @4092. I would have to redo UI to use BinaryInfo classes and merge it in #4092. 

We would need to implement different plugins to support download, github download, brew and choc update/install but it should fit nicely in that nice API and implementation developed by @axel7083.

@benoitf WDYT?

**Comment by benoitf** [comment]
yes there is a dependency. But I think we need a clear separation. One is about to fetch, list, get notifications from 3rd party repositories

Here it's to provide CLI information, lifecycle, updates.

File: packages/extension-api/src/extension-api.d.ts
Comment (dgolovin) [reply packages/extension-api/src/extension-api.d.ts:2437]: It is here https://github.com/containers/podman-desktop/blob/312dcd2bf988f001c52c6ae2ff73a0cf1c1deeca/packages/extension-api/src/extension-api.d.ts#L316

**Comment by dgolovin** [comment]
@benoitf that is what I have for now. Ideally extensions should declare tools in package.json without any coding.
image

File: packages/extension-api/src/extension-api.d.ts
Comment (dgolovin) [reply packages/extension-api/src/extension-api.d.ts]: I am going to stick to create method returning CliTool, the same pattern used for Providers.

Comment (benoitf) [reply packages/extension-api/src/extension-api.d.ts]: @dgolovin providers use a registration mechanism, not a create method

Comment (benoitf) [reply packages/extension-api/src/extension-api.d.ts]: for example registerProvider vs createStatusbarItem

when we provide something, it's a provider (then a register thing), when we ask the engine to create object for us like a StatusBar, a Messagebox, etc it's a create

but here it's a registerCliToolProvider where you provide the (cliToolProvider object)

Comment (benoitf) [reply packages/extension-api/src/extension-api.d.ts]: I'm a bit lost on 
> I am going to stick to create method returning CliTool, the same pattern used for Providers.

but on https://github.com/containers/podman-desktop/pull/4197#issuecomment-1752429026
image

it's using register approach so I would say you'll stick to a register method returning a Disposable object as shown in the image ?

Comment (dgolovin) [reply packages/extension-api/src/extension-api.d.ts]: I meant to use approach similar to provider module here:
https://github.com/containers/podman-desktop/blob/efbe62aef78467ac79644839d69d7a3a9a89c327/packages/extension-api/src/extension-api.d.ts#L426

Right now stripped diagram looks like this:


Currently implemented on UI side:
1. Simple UI
2. UI update on 'extension-loaded' event
3. UI update on 'cli-tool-create' or 'cli-tool-remove'

On main side:
1. Simple registry above
2. Notification on create and dispose of CliTool
3. Generating of the Info objects for UI
4. Cli Tool declaration in package.json through extension

You can ran it from branch and you will see docker-compose in CLI Tools preferences
If you install/uninstall `ghcr.io/dgolovin/podman-desktop-kubectl-ext:latest` you will see kubectl appears and disappears in UI. Compose is using cli module to define cli tool form activation method. kubctl does it in package.json extension manifest.

https://github.com/containers/podman-desktop/assets/620330/0840e5d4-62ac-48b2-bae4-e0ec7fde31b3

File: packages/renderer/src/stores/cli-tools.ts
Comment (benoitf) [inline packages/renderer/src/stores/cli-tools.ts]: I think it's a copy/paste, here we should have like 'cliTools' (it's a display thing but better to use cliTool wording)

File: packages/renderer/src/stores/cli-tools.ts
Comment (benoitf) [inline packages/renderer/src/stores/cli-tools.ts]:
```
 if ('extensions-already-started' === eventName) {
```

File: packages/renderer/src/stores/cli-tools.ts
Comment (benoitf) [inline packages/renderer/src/stores/cli-tools.ts]:
```suggestion
const windowListeners = ['system-ready', 'extensions-already-started'];
```

I think it should track also when an extension is stopped or started

```ts
windowEvents = ['extension-stopped', 'extensions-started'];
```

File: extensions/compose/compose.png
Comment (benoitf) [inline extensions/compose/compose.png:1]: the file should not be part of the implementation of the provider

but also AFAIK it's not the expected icon for compose (compose is an octopus)

File: packages/main/src/plugin/extension-loader.ts
Comment (benoitf) [inline packages/main/src/plugin/extension-loader.ts]: maybe we need to defer this part as it might evolve as I'm not sure about the usage of the extension name and the displayName of the extension.

I mean, probably you want to register multiple CLIS like multiple commands so it might be multiple createCliTool

File: packages/main/src/plugin/cli-tool-registry.ts
Comment (benoitf) [inline packages/main/src/plugin/cli-tool-registry.ts]: I'm wondering why we need to store the extension name and extension display name as you can find details about an extension from its extension Id

File: packages/main/src/plugin/cli-tool-registry.ts
Comment (benoitf) [inline packages/main/src/plugin/cli-tool-registry.ts]: it is confusing that you get as id an internal thing

so it means if you register 'docker-compose' from an extension you don't have a way to find it back ?

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts:2428]: probably missing a list or get methods to list all cli tools registered ?

**Comment by axel7083** [comment]
@benoitf is there still room for #4092 or should I close it ?

**Comment by benoitf** [comment]
@axel7083 there is room for it.

It's not targeting the same goal. 4092 should be about having ways to be notified for updates, etc from repositories (including Github repository)

so extensions could use 4092 when wanting to download data.

**Comment by axel7083** [comment]
> @axel7083 there is room for it.
> 
> It's not targeting the same goal. 4092 should be about having ways to be notified for updates, etc from repositories (including Github repository)
> 
> so extensions could use 4092 when wanting to download data.

Okey, I will wait for this PR to be merged to adapt #4092 based on the new requirements. Keep me updated !

File: packages/renderer/src/stores/cli-tools.ts
Comment (dgolovin) [reply packages/renderer/src/stores/cli-tools.ts]: in current implementation there are specific events for creating and removing cli tools from store. When extension started/stopped the store items are already populated/disposed.

File: packages/main/src/plugin/extension-loader.ts
Comment (dgolovin) [reply packages/main/src/plugin/extension-loader.ts]: Good catch on registering multiple CLIs. I just copy/pasted the code to test the idea and missed the point there could be many CLIs.

File: extensions/compose/compose.png
Comment (dgolovin) [reply extensions/compose/compose.png:1]: where should the icon come from then? Should extension icon be used?

File: packages/main/src/plugin/cli-tool-registry.ts
Comment (dgolovin) [reply packages/main/src/plugin/cli-tool-registry.ts]: it just to pass less data from main to UI, but we sure can get rid of it.

File: packages/main/src/plugin/extension-loader.ts
Comment (dgolovin) [reply packages/main/src/plugin/extension-loader.ts]: I'll remove this code for now.

File: extensions/compose/compose.png
Comment (benoitf) [reply extensions/compose/compose.png:1]: @mairin do we have a dedicated icon for compose cli?

Comment (dgolovin) [reply extensions/compose/compose.png:1]: compose.png removed, I used icon.png instead

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts]: I would remove this part for now as I'm not sure if it's the way to move forward or why we need this.

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts]: remove for now

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts]: should it return CliTool objects or some 'info' view ? (like a view, not the real object)

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts:2452]: why it contains a label + an id there ? and not only the extensionId ?

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (benoitf) [inline packages/main/src/plugin/api/cli-tool-info.ts]: I'm not sure we can have 'brew' or 'chocolate' here (BTW I think you're speaking about chocolatey ?)

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (benoitf) [inline packages/main/src/plugin/api/cli-tool-info.ts]: remove for now probably

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (benoitf) [inline packages/main/src/plugin/api/cli-tool-info.ts]: it's missing extensionId ?

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (benoitf) [inline packages/main/src/plugin/api/cli-tool-info.ts]: is that providedBy is the extensionId ?

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (benoitf) [reply packages/main/src/plugin/api/cli-tool-info.ts]: why do we need a separate field for binary info ?

for example why the version or location is not already in the upper info object ?

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts]: why CLiTool is extending CliToolOptions ?

**Comment by benoitf** [comment]
let's try to have only the simple state being merged. (no detection, etc)

File: packages/extension-api/src/extension-api.d.ts
Comment (dgolovin) [reply packages/extension-api/src/extension-api.d.ts]: I am not sure ether, but the idea is to be able to configure detection without coding, so you can just add it to package.json or call api directly. We can have multiple predefined version parsers that can be referenced by id. Usually version is a sub-command for a cli tool, it also might have options to control output. It could be `--short` that just prints version in a single line (docker-compose) or `-o json` to print json version (kubectl and oc). 

We can create passthrough and json version parser and confiure detection this way in package.json.

for docker-compose

```
detection?: {
    versionOptions: 'version', '--short',
    versionParser: 'passthrough'
}
```

for kubectl 


```
detection?: {
    versionOptions: 'version', '-o', 'json',
    versionParser: 'json',
    versionParserConfig: {
       path: 'clientVersion.gitVersion'
    }
}
```

and so on.

File: packages/extension-api/src/extension-api.d.ts
Comment (dgolovin) [reply packages/extension-api/src/extension-api.d.ts]: Should it return proxy, that would let call lifecycle methods like install (if we decide to put them in CliTool), but block any other modification for non-owner extensions?

Another option is just expose current internal method that returns Infos used in UI. Is that what you getting at?

File: packages/extension-api/src/extension-api.d.ts
Comment (dgolovin) [reply packages/extension-api/src/extension-api.d.ts:2460]: It is a cache to avoid ipc call to main process to get the name

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (dgolovin) [reply packages/main/src/plugin/api/cli-tool-info.ts]: Extension name to avoid call form UI to main thread.

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (dgolovin) [reply packages/main/src/plugin/api/cli-tool-info.ts]: It could be that way, I was overthinking, that we might have multiple versions available form PATH or from binary storage for downloaded files.

File: packages/extension-api/src/extension-api.d.ts
Comment (dgolovin) [reply packages/extension-api/src/extension-api.d.ts]: because the same properties should be on CliTool anyway, I can copy/paste it if that is better.

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [reply packages/extension-api/src/extension-api.d.ts]: for now it's ok as we have mostly 'properties' but yes I would expect that an extension wouldn't be able to call methods of another CLI registered elsewhere

so don't know if proxy works or just different objects that provides only 'properties' (like is there a lifecycle: yes/no or what are the lifecycle methods but there is no direct call possible on these objects)


if it's easier for now, we can remove the list method from the API

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (benoitf) [reply packages/main/src/plugin/api/cli-tool-info.ts]: ok so probably it should be extensionName or extensionInfo ?

in the API we have
```
 extensionInfo: {
      id: string;
      label: string;
    };
```

and here it looks like we have a different structure. I would probably unify (like switch here to extensionInfo as well then ?)

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (benoitf) [inline packages/main/src/plugin/api/cli-tool-info.ts]: I don't think we have that information easily so I would postpone it for the first iteration

File: packages/main/src/plugin/cli-tool-registry.spec.ts
Comment (benoitf) [inline packages/main/src/plugin/cli-tool-registry.spec.ts]: ok tests are missing

File: packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.svelte
Comment (benoitf) [inline packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.svelte]: should we add that it is not registered ?

```suggestion
      title="No CLI tool has been registered"
```

File: packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.svelte
Comment (benoitf) [inline packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.svelte]:
```suggestion
      message="Start an extension that registers a CLI"
```

File: packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.svelte
Comment (benoitf) [inline packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.svelte]:
```suggestion
            class="float-right text-sm">Registered by {cliTool.providedBy}</span>
```

I think provided is misleading as for example I can have kubectl binary being installed by myself but if an extension find it installed globally it would say that it's provided by this extension while it's more "registered" (but not for example managed or provided)

**Review by benoitf — COMMENTED** [review]
overall, LGTM, I added few minor remarks
Of course unit tests need to be added

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [reply packages/extension-api/src/extension-api.d.ts:2452]: agreed we can just proceed with extensionInfo

File: packages/extension-api/src/extension-api.d.ts
Comment (dgolovin) [reply packages/extension-api/src/extension-api.d.ts]: Let's remove it.

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (dgolovin) [reply packages/main/src/plugin/api/cli-tool-info.ts]: I'll remove them.

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (dgolovin) [reply packages/main/src/plugin/api/cli-tool-info.ts]: Let's unify.

File: packages/main/src/plugin/api/cli-tool-info.ts
Comment (dgolovin) [reply packages/main/src/plugin/api/cli-tool-info.ts]: Should I remove location and version form extension-api.ts as well?

Comment (dgolovin) [reply packages/main/src/plugin/api/cli-tool-info.ts]: State is also seems to be not relevant in this iteration. Should it be removed from first iteration?

Comment (dgolovin) [reply packages/main/src/plugin/api/cli-tool-info.ts]: I removed not easily available info and reduced states to the only one we have 'registered'.

File: extensions/compose/src/compose-extension.ts
Comment (cdrage) [inline extensions/compose/src/compose-extension.ts]: Should we add a note like.. "TODO: Remove before merging?" so we don't merge this in by mistake?

File: extensions/compose/src/compose-extension.ts
Comment (cdrage) [inline extensions/compose/src/compose-extension.ts]: `compose-extension.ts` will be removed in the near future after we 100% get compose onboarding in, the reason being is that `compose-extension.ts` is mainly for the compose status bar on the bottom of Podman Desktop which will be removed in favourite for notifications.

Comment (cdrage) [reply extensions/compose/src/compose-extension.ts]: can you move it to `extension.ts`?

you can also re-use the same description for `extension.ts`

Comment (cdrage) [reply extensions/compose/src/compose-extension.ts]: you'll be able to use the same description once https://github.com/containers/podman-desktop/pull/4179 is added since it'll add the compose "provider" to the list.

File: packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.svelte
Comment (cdrage) [inline packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.svelte]: I feel like the below section can be put into a different modal?

it'll also be easier when writing the tests for the .svelte

File: packages/extension-api/src/extension-api.d.ts
Comment (cdrage) [inline packages/extension-api/src/extension-api.d.ts:2442]: Are you able to add comments similar to other parts of extension-api.d.ts? This will help others when implementing their own extension on how to use the CLI functionality and this functionality is regarding creating update / download / management sections for podman desktop.

**Review by cdrage — CHANGES_REQUESTED** [review]
The implementation looks great and testing it against Compose works! Getting the initial functionality is great.

@benoitf should we hide this behind an experimental flag so we don't show the "Cli Tools" section which wouldn't have full functionality yet?

**Comment by dgolovin** [comment]
@benoitf @cdrage we can move UI part into different PR for #3781, because this PR should be about API only. WDYT?

File: packages/extension-api/src/extension-api.d.ts
Comment (dgolovin) [reply packages/extension-api/src/extension-api.d.ts:2442]: done.

**Comment by benoitf** [comment]
I'm ok @dgolovin to split the pr

File: packages/main/src/plugin/cli-tool-registry.spec.ts
Comment (dgolovin) [reply packages/main/src/plugin/cli-tool-registry.spec.ts]: added tests for cli API

File: packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.spec.ts
Comment (benoitf) [inline packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.spec.ts]: probably should not belong to this PR ?

**Review by benoitf — COMMENTED** [review]
AFAIK some files are moved to https://github.com/containers/podman-desktop/pull/4387

so here we should remove renderer/ part ?

File: packages/renderer/src/lib/preferences/PreferencesPage.svelte
Comment (benoitf) [inline packages/renderer/src/lib/preferences/PreferencesPage.svelte]: I suppose we need to remove as well this change ?

**Review by benoitf — APPROVED** [review]
LGTM (removing the renderer change)

File: extensions/compose/src/compose-extension.ts
Comment (dgolovin) [reply extensions/compose/src/compose-extension.ts]: moved to separate PR https://github.com/containers/podman-desktop/pull/4388

File: extensions/compose/src/compose-extension.ts
Comment (dgolovin) [reply extensions/compose/src/compose-extension.ts]: Moved to PR https://github.com/containers/podman-desktop/pull/4388

File: packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.spec.ts
Comment (dgolovin) [reply packages/renderer/src/lib/preferences/PreferencesCliToolsRendering.spec.ts]: removed

**Review by cdrage — APPROVED** [review]
LGTM. This is awesome 🚀  thanks for addressing the comments.

**Pull request merged**
